### PR TITLE
New Reader - State Restoration, Spotlight Search, Universal Links

### PR DESCRIPTION
**Pull request by Gio2018**
Ref #13580

To test:

##### State Restoration

- go to the Reader
- put the app in background
- kill and relaunch
- verify that the app keeps the state it was left in (e.g. you were on the Likes tab)

##### Spotlight search
- visit a post or a site a few times so it gets indexed
- put the app in background, and pull down to show Spotlight search
- tap on any wordPress icon referring to a specific post, site, or reader page and verify that you get redirected to the proper screen in the app (might need to start typing in the Spotlight Search bar)

##### Universal links
- With the simulator running, use the command `xcrun simctl openurl booted https://wordpress.com/[path]` to test universal links.
- Available paths:
  - `/read`
  - `/discover`
  - `/read/search`
  - `/read/a8c` (requires the account to be part of the A8C team)
  - `/activities/likes`
  - `/following/manage` **(not yet implemented)**
  - `/read/list/[username]/[list_name]` (requires that the account has registered at least one list)
  - `/tag/[tag_name]`
  - `/read/feeds/[feedID]` -> Site (type: ReaderSiteTopic)
  - `/read/blogs/[siteID]` -> Blog (type: ReaderSiteTopic)
  - `/read/feeds/[feedID]/posts/[feedItemID]` (feed (site) post) 
  - `/read/blogs/[siteID]/posts/[postID]` (blog post)

PR submission checklist:

- [x] ~~I have considered adding unit tests where possible.~~ addressed in a separate PR.
- [x] ~~I have considered adding accessibility improvements for my changes.~~ addressed in a separate PR.
- [x] ~~I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.~~ Not released

### CODE REVIEW

File: WordPress/Classes/ViewRelated/Reader/Tab Navigation/ReaderTabViewModel.swift [hunk 1]
```diff
@@ -1,34 +1,78 @@
 
-class ReaderTabViewModel {
-
-    var tabSelectionCallback: ((ReaderAbstractTopic?) -> Void)?
-    var selectedFilter: ReaderAbstractTopic?
+@objc class ReaderTabViewModel: NSObject {
+    // MARK: - Properties
+    /// tab selection
```
Comment (frosty) [inline WordPress/Classes/ViewRelated/Reader/Tab Navigation/ReaderTabViewModel.swift:4]: Could these doc comments have a little more detail?

(and perhaps we could introduce some extra newlines in here to group items / space things out a little)

File: WordPress/Classes/ViewRelated/Reader/Tab Navigation/ReaderTabViewModel.swift [hunk 2]
```diff
@@ -89,11 +136,12 @@ extension ReaderTabViewModel {
             guard let topics = try ContextManager.sharedInstance().mainContext.fetch(topicsFetchRequest) as? [ReaderAbstractTopic] else {
                 return
             }
-
-            completion(ReaderHelpers.rearrange(items: topics.map { ReaderTabItem(topic: $0) }))
+            let items = ReaderHelpers.rearrange(items: topics.map { ReaderTabItem(topic: $0) })
```
Comment (frosty) [inline WordPress/Classes/ViewRelated/Reader/Tab Navigation/ReaderTabViewModel.swift:139]: Do we need the extra `items` variable here? Could we just use `tabItems` directly?

File: WordPress/Classes/ViewRelated/Reader/Tab Navigation/WPTabBarController+ReaderTabNavigation.swift [hunk 3]
```diff
@@ -1,9 +1,107 @@
...
+    @objc func makeReaderTabViewModel() -> ReaderTabViewModel {
+        let viewModel = ReaderTabViewModel(readerContentFactory: makeReaderContentViewController(with:),
+                                           searchNavigationFactory: navigateToReaderSearch)
+        return viewModel
+    }
+
+    func makeReaderContentViewController(with topic: ReaderAbstractTopic?) -> ReaderContentViewController {
+        var controller: ReaderStreamViewController
+        if let topic = topic {
+            controller = ReaderStreamViewController.controllerWithTopic(topic)
```
Comment (frosty) [inline WordPress/Classes/ViewRelated/Reader/Tab Navigation/WPTabBarController+ReaderTabNavigation.swift:27]: We could just return these controlls directly instead of assigning to a variable

File: WordPress/Classes/ViewRelated/Reader/Tab Navigation/WPTabBarController+ReaderTabNavigation.swift [hunk 4]
```diff
@@ -1,9 +1,107 @@
...
+    private func makeReaderTabView(_ viewModel: ReaderTabViewModel) -> ReaderTabView {
+        return ReaderTabView(viewModel: self.readerTabViewModel)
+    }
+}
+
+
+/// Reader Navigation
+extension WPTabBarController {
+
+    /// reader navigation methods
```
Comment (frosty) [inline WordPress/Classes/ViewRelated/Reader/Tab Navigation/WPTabBarController+ReaderTabNavigation.swift:43]: These comments might be better as `// MARK: - Navigation` style comments, to group the various methods together.

File: WordPress/Classes/ViewRelated/Reader/Tab Navigation/WPTabBarController+ReaderTabNavigation.swift [hunk 5]
```diff
@@ -1,9 +1,107 @@
...
+    }
+
+    func switchToDiscover() {
+        navigateToReader()
+        readerTabViewModel.switchToTab(where: {
+            ReaderHelpers.topicIsDiscover($0)
+        })
+    }
+
+    func swithcToMyLikes() {
```
Comment (frosty) [inline WordPress/Classes/ViewRelated/Reader/Tab Navigation/WPTabBarController+ReaderTabNavigation.swift:88]: Typo `swithc`

File: WordPress/Classes/ViewRelated/Reader/Tab Navigation/WPTabBarController+ReaderTabNavigation.swift [hunk 6]
```diff
@@ -1,9 +1,107 @@
...
+        navigateToReader()
+        readerTabViewModel.switchToTab(where: {
+            ReaderHelpers.topicIsLiked($0)
+        })
+    }
+
+    /// switches to a manu item topic that satisfies the given predicate with a topic value
```
Comment (frosty) [inline WordPress/Classes/ViewRelated/Reader/Tab Navigation/WPTabBarController+ReaderTabNavigation.swift:95]: Typo here and the next comment `manu`

File: WordPress/Classes/ViewRelated/Reader/Tab Navigation/ReaderTabView.swift [hunk 7]
```diff
@@ -73,16 +80,19 @@ extension ReaderTabView {
         tabBar.addTarget(self, action: #selector(selectedTabDidChange(_:)), for: .valueChanged)
 
         viewModel.fetchReaderMenu() { [weak self] items in
-            guard let items = items, let self = self else {
+            guard let items = items, items.count > 0, let self = self else {
```
Comment (bjtitus) [inline WordPress/Classes/ViewRelated/Reader/Tab Navigation/ReaderTabView.swift:83]: Could probably shorten this to `guard items?.isEmpty() == true` or use `isEmpty()` here.

File: WordPress/Classes/ViewRelated/Reader/Tab Navigation/ReaderTabViewModel.swift (hunk 2 above)
Comment (Gio2018) [reply WordPress/Classes/ViewRelated/Reader/Tab Navigation/ReaderTabViewModel.swift:144]: I'm gonna change this a bit in the next PR, so please bear with me :)

**Comment by Gio2018** [comment]
> 👏
> 
> Everything looks good in my testing except for one small thing with the spotlight opening (which may translate to deep links too, I'm not sure):
> 
> If I'm on an article it looks like a new copy is opened and pushed instead of just staying on the already open article. I tried in a previous app version and it doesn't seem like this happens.

Thanks for this @bjtitus ! Should be fixed now

File: WordPress/Classes/ViewRelated/Reader/Tab Navigation/ReaderTabView.swift (hunk 7 above)
Comment (Gio2018) [reply WordPress/Classes/ViewRelated/Reader/Tab Navigation/ReaderTabView.swift:83]: this is gonna change a bit in the new PR, please bear with me :)
